### PR TITLE
Feature/product1

### PR DESCRIPTION
**Pull request by sauravces**
topic covered
pages inside ssr folder are Server Side Component.
- server actions
- revalidate 
- SSR 
- Zod

### CODE REVIEW

File: __test__/productlist.test.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    expect(screen.getByTestId('loading-spinner')).toBeInTheDocument();
+
+    await waitFor(() => {
+      expect(screen.queryByTestId('loading-spinner')).not.toBeInTheDocument();
+      expect(screen.getByText('Id')).toBeInTheDocument();
+      expect(screen.getByText('Name')).toBeInTheDocument();
+      expect(screen.getByText('Description')).toBeInTheDocument();
+      expect(screen.getByText('Price')).toBeInTheDocument();
+      expect(screen.getByText('Actions')).toBeInTheDocument();
```
Comment (aashikvilla) [inline __test__/productlist.test.tsx:36]: Feels like lot of asserts in a single test, try splitting up test

File: __test__/productlist.test.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+describe('ProductList', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  it('renders product table with data', async () => {
+    const mockProducts = [
+      { id: 1, name: 'Product 1', description: 'Description 1', price: 10 },
+      { id: 2, name: 'Product 2', description: 'Description 2', price: 20 },
```
Comment (aashikvilla) [inline __test__/productlist.test.tsx:21]: use faker for test data

File: __test__/createproduct.test.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    await waitFor(() => {
+      expect(screen.getByText('Name is required')).toBeInTheDocument();
+      expect(screen.getByText('Description is required')).toBeInTheDocument();
+      expect(screen.getByText('Price is required')).toBeInTheDocument();
+    });
+  });
+
+  it('render create product titel',()=>{
+    render(<CreateProduct/>);
+    const title=screen.getByText("Create Product",{exact:false})
```
Comment (aashikvilla) [inline __test__/createproduct.test.tsx:26]: Why do we need this ? -> ,{exact:false}

File: __test__/home.test.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import '@testing-library/jest-dom'
+import { render, screen } from '@testing-library/react'
+import Home from '../src/app/home/page'
+ 
+describe('Page', () => {
+    it('renders a heading of about page', () => {
+        render(<Home />)
+        const heading = screen.getByRole('heading', {
+          name: /Home/i,
```
Comment (aashikvilla) [inline __test__/home.test.tsx:9]: check with case case , instead of invariant

File: src/app/service/request.tsx [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const baseUrl="https://localhost:7046/api";
+export default function useRequests(){
```
Comment (aashikvilla) [inline src/app/service/request.tsx:2]: prefix use is generally meant for hook, avoid this for constants

File: src/app/product/[productId]/page.tsx [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    name: "",
+    description: "",
+    price: null,
+  });
+
+  useEffect(() => {
+    fetchProduct();
+  }, [productId]);
+
+  const fetchProduct = async () => {
```
Comment (aashikvilla) [inline src/app/product/[productId]/page.tsx:46]: try to segregate components and services.
Try fetching most of data only on server components

File: src/app/ssr/addForm.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      });
+    }
+    if (success) {
+      router.push("/ssr/productlist");
+    }
+  };
+
+  return (
+    <>
+      <h1>Server Action Page</h1>
```
Comment (aashikvilla) [inline src/app/ssr/addForm.tsx:46]: Move all text, placeholder, to constants, which can used while testing as well

File: src/app/service/request.tsx [hunk 8]
```diff
@@ -0,0 +1,11 @@
+const baseUrl="https://localhost:7046/api";
```
Comment (aashikvilla) [inline src/app/service/request.tsx:1]: use env file

**Review by aashikvilla — COMMENTED** [review]
Feels like a mix of old and new changes, better creating a new one with best practices on next js.

File: __test__/productlist.test.tsx (hunk 2 above)
Comment (saurav-ces) [reply __test__/productlist.test.tsx:21]: ok sure.

**Pull request merged**